plugins/net_plugin/net_plugin.cpp

```diff
@@ -3821,39 +3736,50 @@ namespace eosio {
             fc_wlog( logger, "bad block connection - ${cid}: #${n} ${id}...: unknown exception",
                      ("cid", cid)("n", ptr->block_num())("id", id.str().substr(8,16)));
          }
-         if( exception ) {
+         if( exception || !obh) {
+            if (!obh) {
+               if (prev_is_proper_svnn_block || !ptr->is_proper_svnn_block()) {
+                  fc_dlog(logger, "unlinkable_block ${bn} : ${id}, previous ${pn} : ${pid}",
+                          ("bn", ptr->block_num())("id", id)("pn", block_header::num_from_id(ptr->previous))("pid", ptr->previous));
+               }
+            }
             c->strand.post( [c, id, blk_num=ptr->block_num(), close_mode]() {
                my_impl->sync_master->rejected_block( c, blk_num, close_mode );
                my_impl->dispatcher.rejected_block( id );
             });
             return;
          }
 
+         assert(obh);
+         uint32_t block_num = obh->block_num();
+         prev_is_proper_svnn_block = obh->header().is_proper_svnn_block();
 
-         uint32_t block_num = obt ? obt->block_num() : 0;
+         fc_dlog( logger, "validated block header, best_head ${bt}, broadcasting immediately, connection - ${cid}, blk num = ${num}, id = ${id}",
+                  ("bt", best_head)("cid", cid)("num", block_num)("id", obh->id()) );
+         my_impl->dispatcher.add_peer_block( obh->id(), cid ); // no need to send back to sender
+         my_impl->dispatcher.bcast_block( obh->block(), obh->id() );
 
-         if( block_num != 0 ) {
-            assert(obt);
-            fc_dlog( logger, "validated block header, broadcasting immediately, connection - ${cid}, blk num = ${num}, id = ${id}",
-                     ("cid", cid)("num", block_num)("id", obt->id()) );
-            my_impl->dispatcher.add_peer_block( obt->id(), cid ); // no need to send back to sender
-            my_impl->dispatcher.bcast_block( obt->block(), obt->id() );
-         }
+         if (best_head) {
+            fc_dlog(logger, "posting block ${n} to app thread", ("n", ptr->block_num()));
+            app().executor().post(priority::medium, exec_queue::read_write,
```

Comment:
I'm not understanding why we are posting this to the app thread?

Reply:
`chain_plugin->accept_block()` calls into the controller to process the block which has to be on the main thread. In Part 2 this only calls controller to apply blocks without a specific block.